### PR TITLE
feat(tasks): introduce tasks/lint_rules

### PR DESCRIPTION
**Pull request by leaysgur**
Part of #2020 

- [x] A new task should be added to tasks/.
- [x] It will pull all our supported plugins from their source,
- [x] compare them against our implemented rules,
- [ ] and generate a markdown file with a tracking list.

<details>
<summary>Current generated list for eslint.all</summary>

```sh
❯ cargo run -p lint_rules
- [ ] accessor-pairs
- [x] array-callback-return
- [ ] arrow-body-style
- [ ] block-scoped-var
- [ ] camelcase
- [ ] capitalized-comments
- [ ] class-methods-use-this
- [ ] complexity
- [ ] consistent-return
- [ ] consistent-this
- [x] constructor-super
- [ ] curly
- [ ] default-case
- [x] default-case-last
- [ ] default-param-last
- [ ] dot-notation
- [x] eqeqeq
- [x] for-direction
- [ ] func-name-matching
- [ ] func-names
- [ ] func-style
- [x] getter-return
- [ ] grouped-accessor-pairs
- [ ] guard-for-in
- [ ] id-denylist
- [ ] id-length
- [ ] id-match
- [ ] init-declarations
- [ ] line-comment-position
- [ ] logical-assignment-operators
- [ ] max-classes-per-file
- [ ] max-depth
- [ ] max-lines
- [ ] max-lines-per-function
- [ ] max-nested-callbacks
- [ ] max-params
- [ ] max-statements
- [ ] multiline-comment-style
- [ ] new-cap
- [ ] no-alert
- [x] no-array-constructor
- [x] no-async-promise-executor
- [ ] no-await-in-loop
- [x] no-bitwise
- [x] no-caller
- [x] no-case-declarations
- [x] no-class-assign
- [x] no-compare-neg-zero
- [x] no-cond-assign
- [x] no-console
- [x] no-const-assign
- [x] no-constant-binary-expression
- [x] no-constant-condition
- [ ] no-constructor-return
- [ ] no-continue
- [x] no-control-regex
- [x] no-debugger
- [x] no-delete-var
- [ ] no-div-regex
- [ ] no-dupe-args
- [x] no-dupe-class-members
- [x] no-dupe-else-if
- [x] no-dupe-keys
- [x] no-duplicate-case
- [ ] no-duplicate-imports
- [ ] no-else-return
- [x] no-empty
- [x] no-empty-character-class
- [ ] no-empty-function
- [x] no-empty-pattern
- [x] no-empty-static-block
- [ ] no-eq-null
- [x] no-eval
- [x] no-ex-assign
- [ ] no-extend-native
- [ ] no-extra-bind
- [x] no-extra-boolean-cast
- [ ] no-extra-label
- [x] no-fallthrough
- [x] no-func-assign
- [x] no-global-assign
- [ ] no-implicit-coercion
- [ ] no-implicit-globals
- [ ] no-implied-eval
- [x] no-import-assign
- [ ] no-inline-comments
- [x] no-inner-declarations
- [ ] no-invalid-regexp
- [ ] no-invalid-this
- [x] no-irregular-whitespace
- [ ] no-iterator
- [ ] no-label-var
- [ ] no-labels
- [ ] no-lone-blocks
- [ ] no-lonely-if
- [ ] no-loop-func
- [x] no-loss-of-precision
- [ ] no-magic-numbers
- [ ] no-misleading-character-class
- [ ] no-multi-assign
- [ ] no-multi-str
- [ ] no-negated-condition
- [ ] no-nested-ternary
- [ ] no-new
- [ ] no-new-func
- [ ] no-new-native-nonconstructor
- [ ] no-new-wrappers
- [ ] no-nonoctal-decimal-escape
- [x] no-obj-calls
- [ ] no-object-constructor
- [ ] no-octal
- [ ] no-octal-escape
- [ ] no-param-reassign
- [ ] no-plusplus
- [ ] no-promise-executor-return
- [ ] no-proto
- [x] no-prototype-builtins
- [x] no-redeclare
- [x] no-regex-spaces
- [ ] no-restricted-exports
- [ ] no-restricted-globals
- [ ] no-restricted-imports
- [ ] no-restricted-properties
- [ ] no-restricted-syntax
- [ ] no-return-assign
- [ ] no-script-url
- [x] no-self-assign
- [x] no-self-compare
- [ ] no-sequences
- [x] no-setter-return
- [ ] no-shadow
- [x] no-shadow-restricted-names
- [x] no-sparse-arrays
- [ ] no-template-curly-in-string
- [ ] no-ternary
- [ ] no-this-before-super
- [ ] no-throw-literal
- [x] no-undef
- [ ] no-undef-init
- [ ] no-undefined
- [ ] no-underscore-dangle
- [ ] no-unexpected-multiline
- [ ] no-unmodified-loop-condition
- [ ] no-unneeded-ternary
- [ ] no-unreachable
- [ ] no-unreachable-loop
- [x] no-unsafe-finally
- [x] no-unsafe-negation
- [x] no-unsafe-optional-chaining
- [ ] no-unused-expressions
- [x] no-unused-labels
- [x] no-unused-private-class-members
- [ ] no-unused-vars
- [ ] no-use-before-define
- [ ] no-useless-assignment
- [ ] no-useless-backreference
- [ ] no-useless-call
- [x] no-useless-catch
- [ ] no-useless-computed-key
- [ ] no-useless-concat
- [ ] no-useless-constructor
- [x] no-useless-escape
- [ ] no-useless-rename
- [ ] no-useless-return
- [x] no-var
- [ ] no-void
- [ ] no-warning-comments
- [ ] no-with
- [ ] object-shorthand
- [ ] one-var
- [ ] operator-assignment
- [ ] prefer-arrow-callback
- [ ] prefer-const
- [ ] prefer-destructuring
- [ ] prefer-exponentiation-operator
- [ ] prefer-named-capture-group
- [ ] prefer-numeric-literals
- [ ] prefer-object-has-own
- [ ] prefer-object-spread
- [ ] prefer-promise-reject-errors
- [ ] prefer-regex-literals
- [ ] prefer-rest-params
- [ ] prefer-spread
- [ ] prefer-template
- [ ] radix
- [ ] require-atomic-updates
- [ ] require-await
- [ ] require-unicode-regexp
- [x] require-yield
- [ ] sort-imports
- [ ] sort-keys
- [ ] sort-vars
- [ ] strict
- [ ] symbol-description
- [ ] unicode-bom
- [x] use-isnan
- [x] valid-typeof
- [ ] vars-on-top
- [ ] yoda
```

</details>

### CODE REVIEW

**Comment by Boshen** [comment]
Looks great! Maybe we want to track against all rules instead of just recommended? Since we have implemented some of the non-recommded rules.

File: tasks/lint_rules/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+// TODOs:
+// - Error handling, message
+//   - JS parsing may fail?
+// - Use Result properly
+// - Support other rules
+// - Better AST traversal... :(
+
+const ORIGINAL: &str = "https://raw.githubusercontent.com/eslint/eslint/main/packages/js/src/configs/eslint-recommended.js";
+const OURS_DIR: &str = "crates/oxc_linter/src/rules/eslint";
```
Comment (Boshen) [inline tasks/lint_rules/src/main.rs:17]: We need to import the linter crate so we can get more metadata

Comment (leaysgur) [reply tasks/lint_rules/src/main.rs:18]: Sorry, could you elaborate on this?

**Comment by leaysgur** [comment]
Updated to use all rules instead of recommended.(Now list is so long! 😮)

**Comment by leaysgur** [comment]
To finalize this PR, I would like to confirm a few things.

1️⃣ Generate a file.

Even as it is now, you can output to any location by piping stdout. (`cmd > ./path/to/file.md`)

If it is better to generate a file in Rust code, where should I output to?

2️⃣ Usage image

We have a number of target plugins.

When this command is called, it should target

- for 1 plugin by specifying name
- for all plugins

Which one do you expect?

3️⃣ How to filter out unsupported rules

Do we need a blacklist?(and how do I know the list?)

**Comment by Boshen** [comment]
Thank you for working on this. Let's merge and iterate. I prefer small PRs :-)

I think the next PR should pull in all the other plugins.

> If it is better to generate a file in Rust code, where should I output to?

What I have in mind is trigger the task in CI and update GitHub issues.

> Which one do you expect?

If we are going to update GitHub issues, it'll be easier with "for 1 plugin by specifying name
".

> Do we need a blacklist?(and how do I know the list?)

I think so, for example we won't be able to implement type checking lint rules. We'll need to hard code it I guess.

**Pull request merged**